[onnx][quantization] Add JIT pass to insert permutes for conv ops

supriyar opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #30799 Add quantized concat conversion
* **#30679 [onnx][quantization] Add JIT pass to insert permutes for conv ops**

Summary:
Caffe2 expects quantized ops to be in NHWC format while pytorch inputs are in NCHW.
Add a jit pass to insert permutes to convert from nchw2nhwc before each conv op and add nhwc2nchw permute after the conv op.
Using graph rewriter to find consecutive redundant permutes and remove them from the graph

Test Plan:
python test/onnx/test_pytorch_onnx_caffe2_quantized.py TestQuantizedOps

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D18790518](https://our.internmc.facebook.com/intern/diff/D18790518)

supriyar committed:
[onnx][quantization] Add JIT pass to insert permutes for conv ops
Summary:
Caffe2 expects quantized ops to be in NHWC format while pytorch inputs are in NCHW.
Add a jit pass to insert permutes to convert from nchw2nhwc before each conv op and add nhwc2nchw permute after the conv op.
Using graph rewriter to find consecutive redundant permutes and remove them from the graph

Test Plan:
python test/onnx/test_pytorch_onnx_caffe2_quantized.py TestQuantizedOps

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

supriyar committed:
Update on "[onnx][quantization] Add JIT pass to insert permutes for conv ops"


Summary:
Caffe2 expects quantized ops to be in NHWC format while pytorch inputs are in NCHW.
Add a jit pass to insert permutes to convert from nchw2nhwc before each conv op and add nhwc2nchw permute after the conv op.
Using graph rewriter to find consecutive redundant permutes and remove them from the graph

Test Plan:
python test/onnx/test_pytorch_onnx_caffe2_quantized.py TestQuantizedOps

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

supriyar committed:
Update on "[onnx][quantization] Add JIT pass to insert permutes for conv ops"


Summary:
Caffe2 expects quantized ops to be in NHWC format while pytorch inputs are in NCHW.
Add a jit pass to insert permutes to convert from nchw2nhwc before each conv op and add nhwc2nchw permute after the conv op.
Using graph rewriter to find consecutive redundant permutes and remove them from the graph

Test Plan:
python test/onnx/test_pytorch_onnx_caffe2_quantized.py TestQuantizedOps

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D18790518](https://our.internmc.facebook.com/intern/diff/D18790518)

[ghstack-poisoned]

supriyar committed:
Update on "[onnx][quantization] Add JIT pass to insert permutes for conv ops"


Summary:
Caffe2 expects quantized ops to be in NHWC format while pytorch inputs are in NCHW.
Add a jit pass to insert permutes to convert from nchw2nhwc before each conv op and add nhwc2nchw permute after the conv op.
Using graph rewriter to find consecutive redundant permutes and remove them from the graph

Test Plan:
python test/onnx/test_pytorch_onnx_caffe2_quantized.py TestQuantizedOps

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D18790518](https://our.internmc.facebook.com/intern/diff/D18790518)

[ghstack-poisoned]

houseroad commented on torch/csrc/jit/passes/onnx/unpack_quantized_weights.cpp:
```diff
@@ -72,18 +72,20 @@ double getScaleFromInput(Node* input_node) {
-    return getScaleFromInput(input_node->inputs()[0]->node());
-  } else if (input_name == "aten::split_with_sizes") {
+  else if (
+      input_name == "quantized::max_pool2d" || input_name == "aten::relu" ||
```
Can we create a white list (i.e., unordered_set) and check input_name exists in the white list or not?

houseroad commented on torch/onnx/utils.py:
```diff
@@ -127,8 +127,22 @@ def _optimize_graph(graph, operator_export_type, _disable_torch_constant_prop=Fa
             torch._C._jit_pass_onnx_unpack_quantized_weights(graph, params_dict)
 
+            # Insert permutes before and after each conv op to ensure correct order.
+            torch._C._jit_pass_onnx_quantization_insert_permutes(graph, params_dict)
```
Is it possible we can use _jit_pass_custom_pattern_based_rewrite_graph to insert permutes as well?

supriyar replied:
I did try that but got an assert of the form 
```
terminate called after throwing an instance of 'c10::Error'
  what():  n->owningGraph() == this && n->inBlockList() INTERNAL ASSERT FAILED at ../torch/csrc/jit/ir.h:1191
```
I suspect adding new ops to the graph isn't updating all the required dependencies correctly.

houseroad commented on torch/csrc/jit/passes/onnx/unpack_quantized_weights.cpp:
```diff
@@ -248,5 +250,55 @@ void UnpackQuantizedWeights(
+    auto op_node = match_vmap.at(vmap.at("r"))->node();
+    auto input_node = match_vmap.at(vmap.at("r"))->node()->inputs()[0]->node();
+
+    Node* permute_node = graph->create(
```
Nit: permute_node_before

houseroad commented on torch/csrc/jit/passes/onnx/unpack_quantized_weights.cpp:
```diff
@@ -248,5 +250,55 @@ void UnpackQuantizedWeights(
+    op_node->insertInput(0, permute_node->output());
+
+    Node* permute_node_after = graph->create(
+        Symbol::fromQualString("quantized::nhwc2nchw"), {input_node->output()});
```
why use input_node->output()? shall we use op_node->outputs()[0] here?

houseroad reviewed (commented):
Overall the PR is good. Some inline comments to address.

supriyar committed:
Update on "[onnx][quantization] Add JIT pass to insert permutes for conv ops"


Summary:
Caffe2 expects quantized ops to be in NHWC format while pytorch inputs are in NCHW.
Add a jit pass to insert permutes to convert from nchw2nhwc before each conv op and add nhwc2nchw permute after the conv op.
Using graph rewriter to find consecutive redundant permutes and remove them from the graph

Test Plan:
python test/onnx/test_pytorch_onnx_caffe2_quantized.py TestQuantizedOps

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D18790518](https://our.internmc.facebook.com/intern/diff/D18790518)

[ghstack-poisoned]

houseroad reviewed (approved):
Looks good. Ship it.
